src/wp-includes/media.php

```diff
@@ -5651,104 +5624,154 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 			}
-		}
-
-		return $postprocess( $loading_attrs, true );
-	}
-
-	// An image with `fetchpriority="high"` cannot be assigned `loading="lazy"` at the same time.
-	if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
-		return $postprocess( $loading_attrs, true );
 	}
 
 	/*
-	 * Do not lazy-load images in the header block template part, as they are likely above the fold.
-	 * For classic themes, this is handled in the condition below using the 'get_header' action.
+	 * The key function logic starts here.
 	 */
-	$header_area = WP_TEMPLATE_PART_AREA_HEADER;
-	if ( "template_part_{$header_area}" === $context ) {
-		// Increase media count if there are images in header above a certian minimum size threshold.
-		$maybe_increase_content_media_count();
-		return $postprocess( $loading_attrs, true );
-	}
-
-	// The custom header image is always expected to be in the header.
-	if ( 'get_header_image_tag' === $context ) {
-		// Increase media count if there are images in header above a certian minimum size threshold.
-		$maybe_increase_content_media_count();
-		return $postprocess( $loading_attrs, true );
-	}
+	$in_viewport          = null;
+	$maybe_increase_count = false;
 
-	// Special handling for programmatically created image tags.
-	if ( 'the_post_thumbnail' === $context || 'wp_get_attachment_image' === $context || 'widget_media_image' === $context ) {
+	// Logic to handle a `loading` attribute that is already provided.
+	if ( isset( $attr['loading'] ) ) {
 		/*
-		 * Skip programmatically created images within post content as they need to be handled together with the other
-		 * images within the post content.
-		 * Without this clause, they would already be considered below which skews the image count and can result in
-		 * the first post content image being lazy-loaded or an image further down the page being marked as a high
-		 * priority.
+		 * Interpret "lazy" as not in viewport. Any other value can be
+		 * interpreted as in viewport (realistically only "eager" or `false`
+		 * to force-omit the attribute are other potential values).
 		 */
-		if ( doing_filter( 'the_content' ) ) {
-			return $loading_attrs;
+		if ( 'lazy' === $attr['loading'] ) {
+			$in_viewport = false;
+		} else {
+			$in_viewport = true;
 		}
+	}
 
-		// Conditionally skip lazy-loading on images before the loop.
-		if (
-			// Only apply for main query but before the loop.
-			$wp_query->before_loop && $wp_query->is_main_query()
+	// Logic to handle a `fetchpriority` attribute that is already provided.
+	if ( isset( $attr['fetchpriority'] ) && 'high' === $attr['fetchpriority'] ) {
+		/*
+		 * If the image was already determined to not be in the viewport (e.g.
+		 * from an already provided `loading` attribute), trigger a warning.
+		 * Otherwise, the value can be interpreted as in viewport, since only
+		 * the most important in-viewport image should have `fetchpriority` set
+		 * to "high".
+		 */
+		if ( false === $in_viewport ) {
```

Comment:
That's a fair point, though I am unsure about a good name. I think your suggestions aren't better because this is not just about lazy-loading, but also about fetchpriority. But I definitely agree with you that `$in_viewport` can also be misleading.

I think we have to make a tradeoff between clarity and concise naming. Really, what this is is "Based on our detection, is this image likely to be in the viewport?". Not sure how to express that in a variable name that is clear yet not super long. Any other ideas? 🤔

Reply:
While trying to deduce the location in the viewport is usually the reason we're prioritizing an image, it's more an assumption than a direct relationship. Just to brainstorm a bit...in reality, what we're trying to do here is indicate that an image is a candidate to be treated with priority by applying `fetchpriority` or avoiding lazy-loading. Perhaps something like: `$is_prioritized`, `$maybe_prioritize`, etc? I want to be mindful of not conflating the word `prioritized` with `fetchpriority` either, so up for other ideas.

Reply:
The variable isn't only there though to handle `fetchpriority`, it is primarily about whether to lazy-load or not. So focusing on priority here would be only a subset of what the variable aims to do. I would argue that `$maybe_in_viewport` would then be a better alternative. While we can't detect it, the point is to make assumptions on whether the image is in the viewport. If we were implementing this on the client-side, that's what this variable would signify, "in viewport". Whether an element is "in viewport" is equivalent with "maybe prioritize", but I'd argue the "in viewport" is then still closer to the primary intent - for example, that variable would have already been needed even if the function was only supporting `loading="lazy"`, which is purely based on whether the element is in viewport or not.

Reply:
> The variable isn't only there though to handle `fetchpriority`

Yes, I know. That's what I was getting at when I said I wanted to be mindful of not conflating the two 😄. `$mabye_in_viewport` is probably fine for now, though I'd still prefer to avoid using viewport at all. To avoid bikeshedding, let's go with that for now and we can revisit in the future.

Reply:
Do you still plan to make this variable name update before merging?

Reply:
@joemcgill Good catch, forgot about it last time. I just pushed the change in https://github.com/WordPress/wordpress-develop/pull/4912/commits/098990a187d5103a27501b8add2180a02e2e76c4.